File: Tutorial_Sample_Code/ReportDocObjectModel/ParametersSubrpt/CS/CS_Win_RDObjMod_ParametersSubrpt/Form1.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

using CrystalDecisions.CrystalReports.Engine;
using CrystalDecisions.Shared;


namespace CS_Win_RDObjMod_ParametersSubrpt
{
	public partial class Form1 : Form
	{
		private ReportDocument customersByCityReport;
		private const string PARAMETER_FIELD_NAME = "City";
		private const string SUBREPORT_PARAMETER_FIELD_NAME = "OrderDateRange";
		private const string SUBREPORT_NAME = "CustomerOrders";

		public Form1()
		{
			InitializeComponent();
		}

		private void ConfigureCrystalReports()
		{
			customersByCityReport = new ReportDocument();
			string reportPath = Application.StartupPath + "\\" + "CustomersByCity.rpt";
			customersByCityReport.Load(reportPath);

			ArrayList arrayList = new ArrayList();
			arrayList.Add("Paris");
			arrayList.Add("Tokyo");

			string startDate = "8/1/1997";
			string endDate = "8/31/1997";
			SetDateRangeForOrders(customersByCityReport, startDate, endDate);

			defaultParameterValuesList.DataSource = GetDefaultValuesFromParameterField(customersByCityReport);

			crystalReportViewer.ReportSource = customersByCityReport;
			SetCurrentValuesForParameterField(customersByCityReport, arrayList);
		}

		private void Form1_Load(object sender, EventArgs e)
		{
			ConfigureCrystalReports();
		}

		private void SetCurrentValuesForParameterField(ReportDocument reportDocument, ArrayList arrayList)
		{
			ParameterValues currentParameterValues = new ParameterValues();
			foreach (object submittedValue in arrayList)
			{
				ParameterDiscreteValue parameterDiscreteValue = new ParameterDiscreteValue();
				parameterDiscreteValue.Value = submittedValue.ToString();
				currentParameterValues.Add(parameterDiscreteValue);

			}
			ParameterFieldDefinitions parameterFieldDefinitions = reportDocument.DataDefinition.ParameterFields;
			ParameterFieldDefinition parameterFieldDefinition = parameterFieldDefinitions[PARAMETER_FIELD_NAME];
			parameterFieldDefinition.ApplyCurrentValues(currentParameterValues);
		}

		private ArrayList GetDefaultValuesFromParameterField(ReportDocument reportDocument)
		{
			ParameterFieldDefinitions parameterFieldDefinitions = reportDocument.DataDefinition.ParameterFields;
			ParameterFieldDefinition parameterFieldDefinition = parameterFieldDefinitions[PARAMETER_FIELD_NAME];
			ParameterValues defaultParameterValues = parameterFieldDefinition.DefaultValues;
			ArrayList arrayList = new ArrayList();
			
			foreach (ParameterValue parameterValue in defaultParameterValues)
			{
				if (!parameterValue.IsRange)
				{
					ParameterDiscreteValue parameterDiscreteValue = (ParameterDiscreteValue)parameterValue;
					arrayList.Add(parameterDiscreteValue.Value.ToString());
				}
				
			}
			return arrayList;
		}

		private void SetDateRangeForOrders(ReportDocument reportDocument, string startDate, string endDate)
		{
			ParameterRangeValue parameterRangeValue = new ParameterRangeValue();
			parameterRangeValue.StartValue = startDate;
			parameterRangeValue.EndValue = endDate;
			parameterRangeValue.LowerBoundType = RangeBoundType.BoundInclusive;
			parameterRangeValue.UpperBoundType = RangeBoundType.BoundInclusive;
			ParameterFieldDefinitions parameterFieldDefinitions = reportDocument.DataDefinition.ParameterFields;
			ParameterFieldDefinition parameterFieldDefinition = parameterFieldDefinitions[SUBREPORT_PARAMETER_FIELD_NAME, SUBREPORT_NAME];
			parameterFieldDefinition.CurrentValues.Clear();
			parameterFieldDefinition.CurrentValues.Add(parameterRangeValue);
			parameterFieldDefinition.ApplyCurrentValues(parameterFieldDefinition.CurrentValues);
		}


		private void redisplay_Click(object sender, EventArgs e)
		{
			ArrayList arrayList = new ArrayList();

			foreach (string item in defaultParameterValuesList.SelectedItems)
			{
				arrayList.Add(item);
			}

			SetCurrentValuesForParameterField(customersByCityReport, arrayList);

			string startDate = orderStartDate.Text;
			string endDate = orderEndDate.Text;
			SetDateRangeForOrders(customersByCityReport, startDate, endDate);

			crystalReportViewer.ReportSource = customersByCityReport;
		}

	}
}

File: Tutorial_Sample_Code/ReportDocObjectModel/ParametersSubrpt/VB/VB_Win_RDObjMod_ParametersSubrpt/Form1.vb
Imports System.Collections
Imports System.Web.UI.WebControls
Imports CrystalDecisions.CrystalReports.Engine
Imports CrystalDecisions.Shared

Public Class Form1
    Private customersByCityReport As ReportDocument
    Private Const PARAMETER_FIELD_NAME As String = "City"
    Private Const SUBREPORT_PARAMETER_FIELD_NAME As String = "OrderDateRange"
    Private Const SUBREPORT_NAME As String = "CustomerOrders"


    Private Sub ConfigureCrystalReports()
        customersByCityReport = New ReportDocument()
        Dim reportPath As String = Application.StartupPath & "\" & "CustomersByCity.rpt"
        customersByCityReport.Load(reportPath)

        Dim myArrayList As ArrayList = New ArrayList()
        myArrayList.Add("Paris")
        myArrayList.Add("Tokyo")

        Dim startDate As String = "8/1/1997"
        Dim endDate As String = "8/31/1997"
        SetDateRangeForOrders(customersByCityReport, startDate, endDate)

        defaultParameterValuesList.DataSource = GetDefaultValuesFromParameterField(customersByCityReport)

        myCrystalReportViewer.ReportSource = customersByCityReport
        SetCurrentValuesForParameterField(customersByCityReport, myArrayList)
    End Sub

    Private Sub Form1_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        ConfigureCrystalReports()
    End Sub

    Private Sub SetCurrentValuesForParameterField(ByVal myReportDocument As ReportDocument, ByVal myArrayList As ArrayList)
        Dim currentParameterValues As ParameterValues = New ParameterValues()
        For Each submittedValue As Object In myArrayList
            Dim myParameterDiscreteValue As ParameterDiscreteValue = New ParameterDiscreteValue()
            myParameterDiscreteValue.Value = submittedValue.ToString()
            currentParameterValues.Add(myParameterDiscreteValue)
        Next

        Dim myParameterFieldDefinitions As ParameterFieldDefinitions = myReportDocument.DataDefinition.ParameterFields
        Dim myParameterFieldDefinition As ParameterFieldDefinition = myParameterFieldDefinitions(PARAMETER_FIELD_NAME)
        myParameterFieldDefinition.ApplyCurrentValues(currentParameterValues)
    End Sub

    Private Function GetDefaultValuesFromParameterField(ByVal myReportDocument As ReportDocument) As ArrayList
        Dim myParameterFieldDefinitions As ParameterFieldDefinitions = myReportDocument.DataDefinition.ParameterFields
        Dim myParameterFieldDefinition As ParameterFieldDefinition = myParameterFieldDefinitions(PARAMETER_FIELD_NAME)
        Dim defaultParameterValues As ParameterValues = myParameterFieldDefinition.DefaultValues
        Dim myArrayList As ArrayList = New ArrayList()

        For Each myParameterValue As ParameterValue In defaultParameterValues
            If (Not myParameterValue.IsRange) Then
                Dim myParameterDiscreteValue As ParameterDiscreteValue = CType(myParameterValue, ParameterDiscreteValue)
                myArrayList.Add(myParameterDiscreteValue.Value.ToString())
            End If
        Next

        Return myArrayList
    End Function

    Private Sub SetDateRangeForOrders(ByVal myReportDocument As ReportDocument, ByVal startDate As String, ByVal endDate As String)
        Dim myParameterRangeValue As ParameterRangeValue = New ParameterRangeValue()
        myParameterRangeValue.StartValue = startDate
        myParameterRangeValue.EndValue = endDate
        myParameterRangeValue.LowerBoundType = RangeBoundType.BoundInclusive
        myParameterRangeValue.UpperBoundType = RangeBoundType.BoundInclusive

        Dim myParameterFieldDefinitions As ParameterFieldDefinitions = myReportDocument.DataDefinition.ParameterFields

        Dim myParameterFieldDefinition As ParameterFieldDefinition

        Try
            myParameterFieldDefinition = myParameterFieldDefinitions(SUBREPORT_PARAMETER_FIELD_NAME, SUBREPORT_NAME)
            myParameterFieldDefinition.CurrentValues.Clear()
            myParameterFieldDefinition.CurrentValues.Add(myParameterRangeValue)
            myParameterFieldDefinition.ApplyCurrentValues(myParameterFieldDefinition.CurrentValues)
        Catch ex As Exception
        End Try



    End Sub

    Private Sub redisplay_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles redisplay.Click
        Dim myArrayList As ArrayList = New ArrayList()
        For Each item As String In defaultParameterValuesList.SelectedItems
            myArrayList.Add(item)
        Next

        SetCurrentValuesForParameterField(customersByCityReport, myArrayList)

        Dim startDate As String = orderStartDate.Text
        Dim endDate As String = orderEndDate.Text
        SetDateRangeForOrders(customersByCityReport, startDate, endDate)

        myCrystalReportViewer.ReportSource = customersByCityReport
    End Sub

End Class
